zerver/lib/message.py

```diff
@@ -524,3 +527,116 @@ def apply_unread_message_event(state, message):
 
     if message.get('is_mentioned'):
         state['mentions'].append(message_id)
+
+def find_and_store_to_insert_stream_msgs(user_profile,
+                                         all_stream_messages,
+                                         all_stream_subscription_logs,
+                                         all_messages_to_insert):
+    # type: (UserProfile, DefaultDict[int, List[Message]], DefaultDict[int, List[RealmAuditLog]], List[UserMessage]) -> None
+    def store_user_message_to_insert(message):
+        # type: (Message) -> None
+        message = UserMessage(user_profile=user_profile,
+                              message_id=message['id'], flags=0)
+        all_messages_to_insert.append(message)
+
+    for (stream_id, stream_messages) in all_stream_messages.items():
+        stream_subscription_logs = all_stream_subscription_logs[stream_id]
+
+        for log_entry in stream_subscription_logs:
+            if len(stream_messages) > 0:
+                if log_entry.event_type == 'subscription_deactivated':
+                    for stream_message in stream_messages:
+                        if stream_message['id'] <= log_entry.event_last_message_id:
+                            store_user_message_to_insert(stream_message)
+                        else:
+                            break
```

Comment:
This has a lot of nesting, which makes it really hard to read.  At least the top part:

```
            if len(stream_messages) > 0:
```

would be better written inverted as this:

```
if len(stream_messsages) == 0:
    continue
```

to avoid that.  Which actually, I think you could pull to the outer loop, even, unless I'm misreading this?  

That sort of refactoring could make this function a lot more readable.